server.go

```diff
@@ -1607,104 +1563,83 @@ func (s *Server) WriteSeries(database, retentionPolicy string, points []Point) (
 		retentionPolicy = rp.Name
 	}
 
-	// Collect responses for each channel.
-	type resp struct {
-		index uint64
-		err   error
+	// Ensure all required Series and Measurement Fields are created cluster-wide.
+	if err := s.createMeasurementsIfNotExists(database, retentionPolicy, points); err != nil {
+		return 0, err
 	}
-	ch := make(chan resp, len(points))
 
-	// Write each point in parallel.
-	var wg sync.WaitGroup
-	for i := range points {
-		wg.Add(1)
-		go func(p *Point) {
-			index, err := s.writePoint(database, retentionPolicy, p)
-			ch <- resp{index, err}
-			wg.Done()
-		}(&points[i])
+	// Ensure all the required shard groups exist. TODO: this should be done async.
+	if err := s.createShardGroupsIfNotExists(database, retentionPolicy, points); err != nil {
+		return 0, err
 	}
-	wg.Wait()
-	close(ch)
 
-	// Calculate max index and check for errors.
-	var index uint64
-	var err error
-	for resp := range ch {
-		if resp.index > index {
-			index = resp.index
-		}
-		if err == nil && resp.err != nil {
-			err = resp.err
+	// Build writeRawSeriesMessageType publish commands.
+	shardData := make(map[uint64][]byte, 0)
+	for _, p := range points {
+		// Local function makes lock management foolproof.
+		measurement, series, err := func() (*Measurement, *Series, error) {
+			s.mu.RLock()
+			defer s.mu.RUnlock()
+			db := s.databases[database]
+			if db == nil {
+				return nil, nil, ErrDatabaseNotFound
+			}
+			if measurement, series := db.MeasurementAndSeries(p.Name, p.Tags); series != nil {
+				return measurement, series, nil
+			}
+			return nil, nil, ErrSeriesNotFound
+		}()
+		if err != nil {
+			return 0, err
 		}
-	}
-	return index, err
-}
-
-func (s *Server) writePoint(database, retentionPolicy string, point *Point) (uint64, error) {
-	measurement, tags, timestamp, values := point.Name, point.Tags, point.Timestamp, point.Values
-
-	// Sanity-check the data point.
-	if measurement == "" {
-		return 0, ErrMeasurementNameRequired
-	}
-	if len(values) == 0 {
-		return 0, ErrValuesRequired
-	}
 
-	// Find the id for the series and tagset
-	seriesID, err := s.createSeriesIfNotExists(database, measurement, tags)
-	if err != nil {
-		return 0, err
-	}
-
-	// Retrieve measurement.
-	m, err := s.measurement(database, measurement)
-	if err != nil {
-		return 0, err
-	} else if m == nil {
-		return 0, ErrMeasurementNotFound
-	}
+		// Retrieve shard group.
+		g, err := s.shardGroupByTimestamp(database, retentionPolicy, p.Timestamp)
+		if err != nil {
+			return 0, err
+		}
 
-	// Retrieve shard group.
-	g, err := s.createShardGroupIfNotExists(database, retentionPolicy, timestamp)
-	if err != nil {
-		return 0, fmt.Errorf("create shard(%s/%s): %s", retentionPolicy, timestamp.Format(time.RFC3339Nano), err)
-	}
+		// Find appropriate shard within the shard group.
+		sh := g.ShardBySeriesID(series.ID)
 
-	// Find appropriate shard within the shard group.
-	sh := g.ShardBySeriesID(seriesID)
+		// Get a field codec.
+		s.mu.RLock()
```

Comment:
same to this, if you pull the rlock outside the points loop, you'll only have to do it once